assets/js/base/hooks/use-collection-data.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * External dependencies
+ */
+import { useState, useEffect, useMemo } from '@wordpress/element';
+import {
+	useQueryStateByContext,
+	useQueryStateByKey,
+	useCollection,
+} from '@woocommerce/base-hooks';
+import { useCollectionDataContext } from '@woocommerce/base-context/collection-data-context';
+import { useDebounce } from 'use-debounce';
+import { find, sortBy } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { useShallowEqual } from './use-shallow-equal';
+
+const buildCollectionDataQuery = ( collectionDataQueryState ) => {
+	const query = collectionDataQueryState;
+
+	if ( collectionDataQueryState.calculate_attribute_counts ) {
+		query.calculate_attribute_counts = sortBy(
+			collectionDataQueryState.calculate_attribute_counts.map(
+				( { taxonomy, queryType } ) => {
+					return {
+						taxonomy,
+						query_type: queryType,
+					};
+				}
+			),
+			[ 'taxonomy', 'query_type' ]
+		);
+	}
+
+	return query;
+};
+
+export const useCollectionData = ( {
+	queryAttribute,
+	queryPrices,
+	queryState,
+	context,
+} ) => {
+	const collectionDataContext = useCollectionDataContext();
+	context = context || collectionDataContext;
+
+	const [ collectionDataQueryState ] = useQueryStateByContext( context );
+	const [
+		calculateAttributesQueryState,
+		setCalculateAttributesQueryState,
+	] = useQueryStateByKey( 'calculate_attribute_counts', [], context );
+	const [
+		calculatePriceRangeQueryState,
+		setCalculatePriceRangeQueryState,
+	] = useQueryStateByKey( 'calculate_price_range', false, context );
+
+	const currentQueryAttribute = useShallowEqual( queryAttribute || {} );
+	const currentQueryPrices = useShallowEqual( queryPrices );
+
+	useEffect( () => {
+		if ( currentQueryAttribute ) {
+			const foundAttribute = find(
+				calculateAttributesQueryState,
+				( attribute ) => {
+					return (
+						attribute.taxonomy === currentQueryAttribute.taxonomy
+					);
+				}
+			);
+
+			if ( ! foundAttribute ) {
+				const setState = calculateAttributesQueryState;
```

Comment:
This assignment doesn't clone `calculateAttributesQueryState`, I think it might bring confusion. What about rewritting these three lines like this?
```ES6
				setCalculateAttributesQueryState(
					calculateAttributesQueryState.concat( currentQueryAttribute )
				);
```